java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/SymbolicValue.java

```diff
@@ -115,49 +130,40 @@ public void computedFrom(List<SymbolicValue> symbolicValues) {
   }
 
   public List<ProgramState> setConstraint(ProgramState programState, ObjectConstraint nullConstraint) {
-    Object data = programState.getConstraint(this);
-    if (data instanceof ObjectConstraint) {
-      ObjectConstraint nc = (ObjectConstraint) data;
-      if (nc.isNull() ^ nullConstraint.isNull()) {
-        // setting null where value is known to be non null or the contrary
-        return ImmutableList.of();
-      } else if (hasAnyStatus(nc) && hasNoStatus(nullConstraint)) {
-        return ImmutableList.of(programState);
+    Constraint cstraint = programState.getConstraint(this, nullConstraint.getClass());
+    if (cstraint == null) {
+      if(nullConstraint.isNull()) {
+        PMap<Class<? extends Constraint>, Constraint> onlyNullConstraint = PCollections.<Class<? extends Constraint>, Constraint>emptyMap()
+          .put(ObjectConstraint.class, ObjectConstraint.NULL);
+        return ImmutableList.of(programState.addConstraints(this, onlyNullConstraint));
+      } else {
+        return ImmutableList.of(programState.addConstraint(this, nullConstraint));
       }
-    }
-    if (data instanceof BooleanConstraint) {
-      return nullConstraint.isNull() ? ImmutableList.of() : ImmutableList.of(programState);
-    }
-    if (data == null || !data.equals(nullConstraint)) {
-      return ImmutableList.of(programState.addConstraint(this, nullConstraint));
+    } else if (cstraint != nullConstraint) {
+      return ImmutableList.of();
     }
     return ImmutableList.of(programState);
   }
 
-  private static boolean hasAnyStatus(ObjectConstraint constraint) {
-    return !hasNoStatus(constraint);
-  }
-
-  private static boolean hasNoStatus(ObjectConstraint constraint) {
-    return constraint.hasStatus(null);
-  }
-
   public List<ProgramState> setConstraint(ProgramState programState, BooleanConstraint booleanConstraint) {
-    Object data = programState.getConstraint(this);
-    // update program state only for a different constraint
-    if (data instanceof BooleanConstraint && !data.equals(booleanConstraint)) {
-      // setting null where value is known to be non null or the contrary
+    Constraint cstraint = programState.getConstraint(this, booleanConstraint.getClass());
```

Comment:
Not sure we actually need a body for that method anymore, as its body is now completely equivalent to what is done in `setConstraint(ProgramState programState, Constraint constraint)`